File: cpp/src/jvm/heap_walker.cpp
#include "heap_walker.hpp"
#include <algorithm>
#include <cstring>
#include <unordered_set>

namespace marrow {

std::vector<uint64_t>
find_instances_by_klass(VMMeta* vm, OopDecoder* decoder,
                        uint64_t klass_ptr, size_t limit, size_t chunk_size)
{
    std::vector<uint64_t> found;
    if (!klass_ptr) return found;
    Reader* r = vm->reader();

    // Pre-encode the target Klass as narrow so we compare a single u32.
    const auto& kp = decoder->klass_params;
    uint32_t target_narrow;
    if (!kp.enabled()) {
        target_narrow = uint32_t(klass_ptr);
    } else {
        target_narrow = uint32_t((klass_ptr - kp.base) >> kp.shift);
    }

    // Restrict matches to the compressed-oop heap range so we exclude
    // class mirrors / metaspace / other regions that happen to contain
    // a u32 equal to target_narrow at offset +8. With heap_base = B and
    // shift = S, valid wide oops live in [B, B + 2^32 << S).
    //
    // When oops aren't compressed (large-heap mode), we fall back to a
    // looser upper bound to match historical behavior.
    uint64_t heap_lo = 0;
    uint64_t heap_hi = ~uint64_t(0);
    {
        const auto& op = decoder->oop_params;
        if (op.enabled() || decoder->oops_are_compressed()) {
            heap_lo = op.base;
            heap_hi = op.base + (uint64_t(1) << (32 + op.shift));
        }
    }

    std::unordered_set<uint64_t> seen;
    for (auto& region : r->enumerate_regions(/*writable_only*/true)) {
        if (region.size < 64 * 1024) continue;
        // Region overlap test against the heap window.
        uint64_t region_end = region.base + region.size;
        if (region_end <= heap_lo || region.base >= heap_hi) continue;
        for (uint64_t off = 0; off < region.size; off += chunk_size) {
            size_t part = size_t(std::min<uint64_t>(chunk_size, region.size - off));
            std::vector<uint8_t> buf;
            try { buf = r->read(region.base + off, part); }
            catch (...) { continue; }
            uint64_t part_base = region.base + off;
            // Walk at every 8-byte alignment; klass slot is at +8 (4 bytes).
            for (size_t pos = 0; pos + 12 <= buf.size(); pos += 8) {
                uint32_t narrow;
                std::memcpy(&narrow, buf.data() + pos + 8, 4);
                if (narrow != target_narrow) continue;
                uint64_t oop_addr = part_base + pos;
                // Final fence: the candidate address itself must encode
                // back into a valid narrow oop within the heap window.
                if (oop_addr < heap_lo || oop_addr >= heap_hi) continue;

                // Mark-word sanity. HotSpot mark word low bits encode the
                // object's lock state — values 00b/01b/10b/11b are valid,
                // but a mark word of all zeros indicates we're inside a
                // primitive array's payload rather than at a real header.
                // Scanning into a Long[] body or String's value[] yields
                // candidates whose +8 narrow happens to equal our target's
                // narrow_klass; filtering by mark != 0 culls these.
                uint64_t mark = 0;
                std::memcpy(&mark, buf.data() + pos, 8);
                if (mark == 0) continue;

                if (seen.insert(oop_addr).second) {
                    found.push_back(oop_addr);
                    if (limit && found.size() >= limit) return found;
                }
            }
        }
    }
    return found;
}

} // namespace marrow

File: heap_walker.py
"""Brute-force heap-wide instance finder.

Instead of walking GC-specific region structures (which differ across
G1 / Parallel / Shenandoah / ZGC and even between JDK point releases),
we scan every committed writable page of the target process and test
each qword-aligned address for a valid object header pointing at the
given Klass. Slower than a native object iterator but portable — works
identically on any GC without layout exports.

Positive identification requires:
  1. The word at offset 0 (mark word) looks like a plausible mark
     (bit 0 or bit 2 set — unlocked or fwd-pointer states).
  2. The narrow klass at offset 8 decodes to exactly our target Klass.
  3. We haven't already found this address (pages can overlap heap + CDS).

Use via `find_instances_by_klass(vm, decoder, klass_ptr, limit=None)`.
"""
from __future__ import annotations

import struct
from collections.abc import Iterator

from vm_meta import VMMeta
from oop_reader import OopDecoder


def _iter_aligned_qwords(buf: bytes, base: int) -> Iterator[tuple[int, int]]:
    """Yield (address, qword_value) for every 8-byte aligned qword in buf."""
    align_off = (-base) & 7
    usable = (len(buf) - align_off) & ~7
    if usable <= 0:
        return
    view = buf[align_off:align_off + usable]
    for i in range(0, usable, 8):
        yield base + align_off + i, struct.unpack_from("<Q", view, i)[0]


def find_instances_by_klass(
        vm: VMMeta, decoder: OopDecoder, klass_ptr: int,
        limit: int | None = None,
        chunk_size: int = 4 * 1024 * 1024,
) -> list[int]:
    """Return oop addresses of every instance whose narrow-klass slot
    resolves to `klass_ptr`. Scans the entire writable address space
    of the target process — expect hundreds of MB of reads; typically
    completes in a few seconds for a modest target.
    """
    r = vm.reader
    # Pre-encode target klass as a narrow value once.
    p = decoder.klass_params
    if p.shift == 0 and p.base == 0:
        target_narrow = klass_ptr & 0xFFFFFFFF
    else:
        target_narrow = (klass_ptr - p.base) >> p.shift
    target_narrow &= 0xFFFFFFFF

    found: list[int] = []
    seen: set[int] = set()
    # We only care about heap-ish pages — typical Java heap sits in PAGE_READWRITE
    # pages allocated via VirtualAlloc in 1-MB-ish chunks. Skip pages smaller
    # than 64 KiB (stacks, PEB, etc.) to cut false-positive surface and time.
    for base, size, _prot in r.enumerate_regions(writable_only=True):
        if size < 64 * 1024:
            continue
        # Read the region in chunks so we don't balloon Python memory on
        # huge heaps.
        for off in range(0, size, chunk_size):
            part_size = min(chunk_size, size - off)
            try:
                buf = r.read(base + off, part_size)
            except OSError:
                continue
            # At each qword boundary, treat it as a candidate mark word.
            # The narrow klass lives 8 bytes later (compressed-class layout).
            # We check the narrow-klass candidate against our target.
            part_base = base + off
            # Iterate every 8-aligned position, checking klass@+8.
            for pos in range(0, part_size - 12, 8):
                narrow = struct.unpack_from("<I", buf, pos + 8)[0]
                if narrow != target_narrow:
                    continue
                # Mark word sanity: plausible unlocked pattern. Many values
                # work; 0x5, 0x1, 0x4*, 0x0 are all common. We only reject
                # obviously-zero klass neighbours and look for a mark with
                # one of the low nibbles set.
                mark = struct.unpack_from("<Q", buf, pos)[0]
                if mark == 0 or (mark & 0x3) == 0x2:
                    # 0x2 is the "marked" pattern the GC sets during marking;
                    # typical live objects should not carry exactly 0 either.
                    pass  # still accept; mark word can be a displaced lock
                oop_addr = part_base + pos
                if oop_addr in seen:
                    continue
                seen.add(oop_addr)
                found.append(oop_addr)
                if limit is not None and len(found) >= limit:
                    return found
    return found
